File: swift/lib/ProtocGenSwift/ProtocGenSwift/Writer.swift
import Foundation

public protocol Writer {
    static func withCapacity(capacity: Int) throws -> Writer
    func toBuffer() throws -> NSData
    func writeTag(tag: Int)
    func writeByte(v: UInt8)
    func writeVarInt(v: Int)
    func writeVarUInt(v: UInt)
    func writeVarUInt64(v: UInt64)
    func writeData(v: NSData)
    func writeBool(v: Bool)
    func writeFloat32(v: Float32)
    func writeFloat64(v: Float64)
    func writeString(v: String)
    func pushTagMap(map: [Int:(String, Bool)])
    func popTagMap()
}

File: java/lib/ProtocGenJava/src/main/java/com/tonicdesign/protocgenjava/Writer.java
package com.tonicdesign.protocgenjava;

import java.util.Map;

public interface Writer {

    class TagMapValue {

        public final String key;
        public final boolean isRepeated;

        public TagMapValue(String key, boolean isRepeated) {
            this.key = key;
            this.isRepeated = isRepeated;
        }
    }

    byte[] toBuffer();
    void writeTag(int tag);
    void writeByte(byte v);
    void writeVarInt(long v);
    void writeVarUInt(long v);
    void writeBool(boolean v);
    void writeFloat32(float v);
    void writeFloat64(double v);
    void writeString(String v);
    void pushTagMap(Map<Integer, TagMapValue> map);
    void popTagMap();
}
